nemo/collections/asr/parts/utils/audio_utils.py

```diff
@@ -81,3 +84,222 @@ def select_channels(signal: npt.NDArray, channel_selector: Optional[ChannelSelec
         raise ValueError(f'Unexpected value for channel_selector ({channel_selector})')
 
     return signal
+
+
+def sinc_unnormalized(x: float) -> float:
+    """Unnormalized sinc.
+    
+    Args:
+        x: input value
+        
+    Returns:
+        Calculates sin(x)/x 
+    """
+    return np.sinc(x / np.pi)
+
+
+def theoretical_coherence(
+    mic_positions: npt.NDArray,
+    sample_rate: float,
+    field: str = 'spherical',
+    fft_length: int = 512,
+    sound_velocity: float = SOUND_VELOCITY,
+) -> npt.NDArray:
+    """Calculate a theoretical coherence matrix for given mic positions and field type.
+    
+    Args:
+        mic_positions: 3D Cartesian coordinates of microphone positions, shape (num_mics, 3)
+        field: string denoting the type of the soundfield
+        sample_rate: sampling rate of the input signal in Hz
+        fft_length: length of the fft in samples
+        sound_velocity: speed of sound in m/s
+    
+    Returns:
+        Calculated coherence with shape (num_subbands, num_mics, num_mics)
+    """
+    assert mic_positions.shape[1] == 3, "Expecting 3D microphone positions"
+    num_mics = mic_positions.shape[0]
+
+    if num_mics < 2:
+        raise ValueError(f'Expecting at least 2 microphones, received {num_mics}')
+
+    num_subbands = fft_length // 2 + 1
+    angular_freq = 2 * np.pi * sample_rate * np.arange(0, num_subbands) / fft_length
+    desired_coherence = np.zeros((num_subbands, num_mics, num_mics))
+
+    mic_distance = squareform(pdist(mic_positions))
+
+    for p in range(num_mics):
+        desired_coherence[:, p, p] = 1.0
+        for q in range(1, num_mics):
```

Comment:
Should this be:
for q in range(p+1, num_mics):

There seem to duplicates in the current loop